Fix query for next/anytime

nilsberg opened the pull request:
From a user perspective, headings cannot be scheduled.
So I changed the query such that either tasks have no project and no heading, or the parent project is not Someday/Upcoming.
The case that a task *does* have a heading and *doesn't* have a project is not valid, so I omitted it.

AlexanderWillner commented:
Thanks, will test

lucbeaulieu commented:
> The case that a task _does_ have a heading and _doesn't_ have a project is not valid, so I omitted it.

Actually those orphan tasks are quite important from a weekly review perspective. Same thing for projects without next step (no active task but active projects). I still haven't found a way to extract those using only a SQL query...

nilsberg commented:
How would you go about to create a task without a project but with a heading?
I mean in Things, not hacking the DB. 

But less assumptions are usually better!

lucbeaulieu commented:
I meant orphan tasks that are not associate with any projects at all. It is indeed impossible as far as I know to associate a task only to a heading without a project, sorry for the confusion.

AlexanderWillner commented:
@lucbeaulieu : to find empty projects:
```
SELECT
	PROJECT.title
FROM
	TMTask as PROJECT
WHERE
	PROJECT.type = 1
GROUP BY
	PROJECT.uuid
HAVING
	(SELECT COUNT(uuid) FROM TMTask WHERE project = PROJECT.uuid) = 0
```

lucbeaulieu commented:
Yes...I was trying to find a way to achieved that without modifying too much the main structure of your sql sequence in the python file because I would like to be able to display in one of the 7 columns (without modifying also the html writing part) these projects and associated link. In fact I really like your visual presentation (compared to my old, and let's be real badly coded KanbanView) and if instead of displaying individual tasks, It could display Projects sorted by due date is available along with task counts in each of them with columns sometimes being Area title instead of Task Tag or columns corresponding to Projects with no active task or Projects in the Someday box that would be really great and something Things is not very good at: providing a complete visual at 20000 feet of everything important on your plate. 

By the way just download the app. Look great! 

I found out that the Backlog and Waiting columns are empty (I have to change my WaitingFor tag to Waiting for the latter). The Upcoming column as essentially only recurring tasks since I follow GTD and almost never asking due date to individual tasks. Today and Inbox are fine but Things is already quite good at doing that...

AlexanderWillner commented:
I added a new column to show empty projects. I'd also like to have the option to switch between task-based and project-based views...

lucbeaulieu commented:
I have tried this to automatically get counts (at least on active projects) but I do not think I get everything correctly (I am missing a few projects) - I am using DB Browser for sqlite app for testing:
SELECT DISTINCT
                TASK.uuid,
                TASK.title,
		TASK.project,
		COUNT(TASK.project) as TNB,
                CASE
                    WHEN AREA.title IS NOT NULL THEN AREA.title
                    WHEN PROJECT.title IS NOT NULL THEN PROJECT.title
                    WHEN HEADING.title IS NOT NULL THEN HEADING.title
                END,
                CASE
                    WHEN AREA.uuid IS NOT NULL THEN AREA.uuid
                    WHEN PROJECT.uuid IS NOT NULL THEN PROJECT.uuid
                END,
                CASE
                    WHEN TASK.recurrenceRule IS NULL
                    THEN date(TASK.dueDate,"unixepoch")
					ELSE NULL
                END,
                date(TASK.creationDate,"unixepoch"),
                date(TASK.userModificationDate,"unixepoch"),
                date(TASK.startDate,"unixepoch"),
                date(TASK.stopDate,"unixepoch")
            FROM
                TMTask AS TASK
            LEFT OUTER JOIN
                TMTask PROJECT ON TASK.project = PROJECT.uuid
            LEFT OUTER JOIN
                TMArea AREA ON TASK.area = AREA.uuid
            LEFT OUTER JOIN
                TMTask HEADING ON TASK.actionGroup = HEADING.uuid
            LEFT OUTER JOIN
                TMTaskTag TAGS ON TASK.uuid = TAGS.tasks
            LEFT OUTER JOIN
                TMTag TAG ON TAGS.tags = TAG.uuid
			WHERE TASK.type IS  NOT 2 AND TASK.trashed IS 0 AND TASK.status IS 0
			GROUP BY PROJECT.title

AlexanderWillner commented:
To get counts (i.e. # tasks / project): 
```
SELECT
	PROJECT.title,
	(SELECT COUNT(uuid) FROM TMTask WHERE project = PROJECT.uuid)
FROM
	TMTask as PROJECT
WHERE
	PROJECT.type = 1
```
To find projects with no tasks, see above or one of my last commits.

lucbeaulieu commented:
Adding AND PROJECT.status = 0 seems to return all active projects! But it also change the overall SQL call architecture from what you have right now (potentially also needing a change in the html writing part also....)

lucbeaulieu commented:
I forgot that the area title is lost in the process if one needs to group/search by area

AlexanderWillner commented:
Might make sense to create an issue at https://github.com/AlexanderWillner/KanbanView/issues to specify exactly what you're looking for.

nilsberg commented:
> I added a new column to show empty projects. I'd also like to have the option to switch between task-based and project-based views...

I really like the "Grooming" column!
I added `OR PROJECT_TASK.{self.IS_SCHEDULED}` to the query, as projects with upcoming tasks are showing up in your version. (I did not submit a pull request because it is so simple, but tests are green this time. Which also tells me, that you did not account for this situation in your tests.)

I also added a check for recurring tasks, however I don't know how to check for paused or stopped recurring tasks, so this could be misleading. Btw I really like your mini query language.

And I am pondering, whether Someday tasks in a project make it missing a next action or not. I'm leaning towards that Someday tasks do not count as a next action.

lucbeaulieu commented:
> I'm leaning towards that Someday tasks do not count as a next action.

I agree. I also made a few proposal for Project Level KanbanView here: https://github.com/AlexanderWillner/KanbanView/issues/6

I know have a fully working KanbanView at the Project Level (20 000 feet) based on Alex coding (instead of my old 2018 one!).

nilsberg commented:
Get only not paused recurring tasks (not sure how ended recurring tasks are represented):

```python
RECURRING_IS_NOT_PAUSED = "instanceCreationPaused = 0"

(PROJECT_TASK.{self.IS_RECURRING} AND PROJECT_TASK.{self.RECURRING_IS_NOT_PAUSED})
```

Again, if a PR helps, I gladly submit it. But this seems simpler to just include by hand.

AlexanderWillner commented:
@nilsberg ```I added OR PROJECT_TASK.{self.IS_SCHEDULED} to the query, as projects with upcoming tasks are showing up in your version.``` - actually, this is on purpose: imho the project should be scheduled when there is no active task anymore (but only scheduled ones)

AlexanderWillner commented:
@lucbeaulieu ``` I also made a few proposal for Project Level KanbanView here: #6``` - thank you, I already implemented a switch at the top right corner to toggle between task and project view. Very good to also look at the projects.

AlexanderWillner commented:
@nilsberg ```Get only not paused recurring tasks (not sure how ended recurring tasks are represented)``` - thanks, in which context should we add this? Smaller GitHub issues would be better than this thread here ;)

nilsberg commented:
I am sorry, you are totally right. I moved my topic to https://github.com/AlexanderWillner/KanbanView/issues/7#issue-602718236
